converter/dicom-cast/src/Microsoft.Health.DicomCast.Core/Features/Worker/FhirTransaction/RetryableFhirTransactionPipeline.cs

```diff
@@ -21,42 +22,35 @@ namespace Microsoft.Health.DicomCast.Core.Features.Worker.FhirTransaction
     /// </summary>
     public class RetryableFhirTransactionPipeline : IFhirTransactionPipeline
     {
-        private const int MaxRetryCount = 3;
-
         private readonly IFhirTransactionPipeline _fhirTransactionPipeline;
         private readonly IExceptionStore _exceptionStore;
         private readonly IAsyncPolicy _retryPolicy;
+        private readonly IAsyncPolicy _timeoutPolicy;
 
-        public RetryableFhirTransactionPipeline(IFhirTransactionPipeline fhirTransactionPipeline, IExceptionStore exceptionStore)
-            : this(
-                  fhirTransactionPipeline,
-                  exceptionStore,
-                  MaxRetryCount)
-        {
-        }
-
-        internal RetryableFhirTransactionPipeline(
-            IFhirTransactionPipeline fhirTransactionPipeline,
-            IExceptionStore exceptionStore,
-            int maxRetryCount)
+        public RetryableFhirTransactionPipeline(IFhirTransactionPipeline fhirTransactionPipeline, IExceptionStore exceptionStore, IOptions<RetryConfiguration> retryConfiguration)
         {
             EnsureArg.IsNotNull(fhirTransactionPipeline, nameof(fhirTransactionPipeline));
             EnsureArg.IsNotNull(exceptionStore, nameof(exceptionStore));
+            EnsureArg.IsNotNull(retryConfiguration, nameof(retryConfiguration));
 
             _fhirTransactionPipeline = fhirTransactionPipeline;
             _exceptionStore = exceptionStore;
+            _timeoutPolicy = Policy.TimeoutAsync(retryConfiguration.Value.TotalRetryDuration);
```

Comment:
I read the code for this, and it's actually much better than I thought. They end up linking a new cancellation token (that will timeout after this value) with the one passed into `ExecuteAsync`, so this is just as good as you performing the linking yourself as discussed yesterday! By default, they also call this timeout strategy "optimistic", which is another way of saying that the underlying delegate actually monitors the token (as opposed to simply having to stop waiting on the task after so many seconds and hoping it cleans itself up)

Reply:
Perfect, thanks for looking into it further!